### ISSUE TITLE
Latest version slows down object listing

### ISSUE THREAD

**Issue by josiasmontag** [minio/minio#10873]
After upgrading to `RELEASE.2020-11-10T21-02-24Z` we noticed that listing operations became slow. The problem immediately resolved after downgrading to `RELEASE.2020-10-28T08-16-50Z` again.

Most likely it is somehow related to the new ListObject caching.

## Expected Behavior
`ListObjects` operations with prefix take ~20ms 

## Current Behavior
`ListObjects` operations sometimes take 40-60s. I would say 50% of the responses do have this problem.

## Possible Solution
Downgrade to `RELEASE.2020-10-28T08-16-50Z`

## Steps to Reproduce (for bugs)

We do have a bucket with ~100,000 objects of all sizes. Files are saved in "folders" with typically only one file per folder/prefix.

```
➜  ~ mc admin info production
●  minio1:9000
   Uptime: 9 minutes 
   Version: 2020-10-28T08:16:50Z
   Network: 4/4 OK 
   Drives: 4/4 OK 

●  minio3:9000
   Uptime: 9 minutes 
   Version: 2020-10-28T08:16:50Z
   Network: 4/4 OK 
   Drives: 4/4 OK 

●  minio4:9000
   Uptime: 9 minutes 
   Version: 2020-10-28T08:16:50Z
   Network: 4/4 OK 
   Drives: 4/4 OK 

●  minio2:9000
   Uptime: 9 minutes 
   Version: 2020-10-28T08:16:50Z
   Network: 4/4 OK 
   Drives: 4/4 OK 

3.9 TiB Used, 4 Buckets, 121,668 Objects
16 drives online, 0 drives offline
```

1.  Upgrade to `RELEASE.2020-11-10T21-02-24Z`
2. List request:

```
➜  ~ mc ls --debug production/tasks/001034c9-a769-4704-b56c-31a15f63d774/
mc: <DEBUG> GET /tasks/?location= HTTP/1.1
Host: XXX
User-Agent: MinIO (darwin; amd64) minio-go/v7.0.6 mc/2020-10-03T02:54:56Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20201111/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20201111T144411Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 131
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 11 Nov 2020 14:44:12 GMT
Server: nginx
Vary: Origin
X-Amz-Bucket-Region: lim
X-Amz-Request-Id: 16467B44B18CB786
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  191.43615ms

mc: <DEBUG> GET /tasks/?delimiter=%2F&encoding-type=url&fetch-owner=true&list-type=2&prefix=001034c9-a769-4704-b56c-31a15f63d774%2F HTTP/1.1
Host: XXX
User-Agent: MinIO (darwin; amd64) minio-go/v7.0.6 mc/2020-10-03T02:54:56Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20201111/lim/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20201111T144412Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 692
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 11 Nov 2020 14:44:55 GMT
Server: nginx
Vary: Origin
X-Amz-Bucket-Region: lim
X-Amz-Request-Id: 16467B44B2A8AB0E
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  43.612644484s
[2020-11-11 12:02:21 CET] 348MiB VTS_01_4.mp4
```

-> Repsonse time of **43s**.

2. Downgrade to `RELEASE.2020-10-28T08-16-50Z`
3. List request:

```
➜  ~ mc ls --debug production/tasks/001034c9-a769-4704-b56c-31a15f63d774/
mc: <DEBUG> GET /tasks/?location= HTTP/1.1
Host: XXX
User-Agent: MinIO (darwin; amd64) minio-go/v7.0.6 mc/2020-10-03T02:54:56Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20201111/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: UNSIGNED-PAYLOAD
X-Amz-Date: 20201111T144749Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 131
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 11 Nov 2020 14:47:49 GMT
Server: nginx
Vary: Origin
X-Amz-Bucket-Region: lim
X-Amz-Request-Id: 16467B775A38C28A
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  177.7077ms

mc: <DEBUG> GET /tasks/?delimiter=%2F&encoding-type=url&fetch-owner=true&list-type=2&prefix=001034c9-a769-4704-b56c-31a15f63d774%2F HTTP/1.1
Host: XXX
User-Agent: MinIO (darwin; amd64) minio-go/v7.0.6 mc/2020-10-03T02:54:56Z
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20201111/lim/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20201111T144749Z
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 200 OK
Content-Length: 692
Accept-Ranges: bytes
Connection: keep-alive
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Date: Wed, 11 Nov 2020 14:47:49 GMT
Server: nginx
Vary: Origin
X-Amz-Bucket-Region: lim
X-Amz-Request-Id: 16467B775B52FCB7
X-Xss-Protection: 1; mode=block

mc: <DEBUG> Response Time:  21.505299ms

[2020-11-11 12:02:21 CET] 348MiB VTS_01_4.mp4
```

-> Repsonse time of **21ms**.

## Context
I didn't notice any error in the MinIO logs.

## Your Environment
Docker Swarm on Debian Buster.

```
# uname -a
Linux storage1 4.19.0-12-amd64 #1 SMP Debian 4.19.152-1 (2020-10-18) x86_64 GNU/Linux
```

Let me know how I can help with debugging this.

**Comment by klauspost** [comment]
@josiasmontag 

Is anything else accessing the cluster at the same time?

Is it possible to record `mc admin trace myminio` before and after the call?

Background: If any other list call is running, and was initiated before this call that would include listing this path it will use the result from that listing instead of asking the disks again.

**Comment by josiasmontag** [comment]
It happened on production, so yes there were definitely other list calls at the same time. The other calls were on other prefixes though.

I was not able to reproduce this locally. It seems it is only happening when there are a lot of parallel requests to the cluster.

**Comment by josiasmontag** [comment]
I logged a slow list object call with `mc admin trace --all --verbose`.

@klauspost Can I send you the log file via email? As it is live data I cannot post the log file publicly.

**Comment by klauspost** [comment]
@josiasmontag Sure. klaus [at] minio [dot] io

**Comment by tomsozolins** [comment]
Same problem here.

**Comment by klauspost** [comment]
I can't really see what I am looking for in your trace. It seems like your listings in `processes` are completing fine. 

```
minio4 [REQUEST s3.ListObjectsV1] 11:44:14.140
minio4 GET /processes?prefix=93abb6d5-32b0-4e31-8961-05cd072cedca%2F
...
minio4 [RESPONSE] [11:44:14.159] [ Duration 19.158ms  ↑ 121 B  ↓ 1.4 KiB ]
```

There are some listings on `tasks/test/` that takes about 30 seconds, but assuming that is the prefix with thousands of sub-prefixes that is expected..

If you would like a deeper analysis of your system performance we offer that as part of our subscription service: https://min.io/pricing

There will be a difference in listing speed with the latest upgrade, but most should end up on the positive side. However for some access patterns there will be some slower scenarios.

**Comment by utking** [comment]
@klauspost perhaps it is expected behavior, but it works fast with 2020-10-28 and terribly slow with 2020-11-06 or newer versions when working with hundreds of objects in a bucket. We also have the same problem, with only several hundred objects under a prefix. The whole picture is - we use distributed Minio (4 hosts, with 1 disk per host) for Harbor (a docker registry). With 2020-10-28, it takes less than 1 seconds to list projects (not even images) on the main Harbor page. With 2020-11-06, loading the projects list takes forever. Even 1 specific project (direct URL) with 5 docker images takes about 20-30 seconds to load. I understand that Harbor is completely separate software, but as soon as you downgrade back to 2020-10-28, the slowness is gone. Even if we forget about using Harbor, but try to access its bucket in Minio, the listing is still slow on all the levels lower than the bucket itself.

**Comment by harshavardhana** [comment]
> There are some listings on `tasks/test/` that takes about 30 seconds, but assuming that is the prefix with thousands of sub-prefixes that is expected..

@utking can you tell us how many objects are present at this prefix?

**Comment by utking** [comment]
The total number for all the buckets
258 GiB Used, 15 Buckets, 138,902 Objects
4 drives online, 0 drives offline
With other buckets much smaller, we can say about 130k objects under the Harbor bucket. They are not piled in the prefix root - the folder structure is usual for Docker registries:
`docker-image-storage/docker/registry/v2/blobs/sha256/...etc`

**Comment by harshavardhana** [comment]
@utking can you try and see how does this help? 

```
MINIO_API_LIST_QUORUM=disk
MINIO_API_EXTEND_LIST_CACHE_LIFE=5m
```

**Comment by utking** [comment]
@harshavardhana , next time I will be able to test it is the upcoming weekend, not earlier, unfortunately.

**Comment by josiasmontag** [comment]
> @utking can you try and see how does this help?
> 
> ```
> MINIO_API_LIST_QUORUM=disk
> MINIO_API_EXTEND_LIST_CACHE_LIFE=5m
> ```

For me these settings do not have any positive effect.

**Comment by harshavardhana** [comment]
> > @utking can you try and see how does this help?
> > ```
> > MINIO_API_LIST_QUORUM=disk
> > MINIO_API_EXTEND_LIST_CACHE_LIFE=5m
> > ```
> 
> For me these settings do not have any positive effect.

Can you provide more info regarding directory heirarchy? etc

**Comment by josiasmontag** [comment]
Bucket `tasks`:
~ 1,000 prefixes
Each prefix has 1-5 objects. No sub-prefixes.

```
tasks/1/file.pdf
tasks/2/file1.pdf
tasks/2/file2.pdf
tasks/2/file3.pdf
....
tasks/1000/file.pdf
```

For example, the prefix `tasks/test/` has 1 object and listing takes 40s (most of the time, sometimes it is fast). Verbose log of this was provided to @klauspost.

**Comment by klauspost** [comment]
As said, if a listing is running that will produce the output as a sideeffect this listing is used instead of initiating a new one.

So if a recursive listing is running on `tasks/` and you concurrently ask for a listing of `tasks/1` the second request will wait for the already running listing to get to `tasks/1` and grab the result from that. That would be true for any sub-path of `tasks/`.

**Comment by harshavardhana** [comment]
> So if a recursive listing is running on `tasks/` and you concurrently ask for a listing of `tasks/1` the second request will wait for the already running listing to get to `tasks/1` and grab the result from that. That would be true for any sub-path of `tasks/`.

@klauspost would it be possible to make it independent? is there a caveat on that?

**Comment by utking** [comment]
@klauspost 
> As said, if a listing is running that will produce the output as a sideeffect this listing is used instead of initiating a new one.

no questions about this. the question is what changed after 2020-10-28 that started causing the slowness.

**Comment by harshavardhana** [comment]
> no questions about this. the question is what changed after 2020-10-28 that started causing the slowness.

We have changed the way listing works, its actually now cached for performance.

**Comment by josiasmontag** [comment]
> So if a recursive listing is running on `tasks/` and you concurrently ask for a listing of `tasks/1` the second request will wait for the already running listing to get to `tasks/1` and grab the result from that. That would be true for any sub-path of `tasks/`.

This insight is interesting. In my specific case there are no (intentional) recursive listings. There might be parallel listings of `tasks/1/` and of `tasks/` with `/` delimiter. But the delimiter is always set, meaning no recursive listing and there should be no reason that the listing gets blocked by the other listing, right?

**Comment by harshavardhana** [comment]
Fixed by #10920

**Comment by hisayan** [comment]
I have encountered a phenomenon similar to this.
Isn't the same thing happening?

minio version RELEASE.2021-11-09T03-21-45Z
mc version RELEASE.2021-11-05T10-05-06Z

**Comment by harshavardhana** [comment]
> I have encountered a phenomenon similar to this. Isn't the same thing happening?
> 
> minio version RELEASE.2021-11-09T03-21-45Z mc version RELEASE.2021-11-05T10-05-06Z

This shouldn't be happening - most probably your setup is poor.  Please open a new issue with some more details - let's not discuss older issues.
